// d-fi/dist/src/lib/signale.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.note = exports.error = exports.success = exports.pending = exports.warn = exports.info = void 0;
const chalk_1 = __importDefault(require("chalk"));
const info = (message) => chalk_1.default.blueBright('ℹ info ') + message;
exports.info = info;
const warn = (message) => chalk_1.default.yellowBright('⚠ warn ') + message;
exports.warn = warn;
const pending = (message) => chalk_1.default.blueBright('● pending ') + message;
exports.pending = pending;
const success = (message) => chalk_1.default.greenBright('✔ success ') + message;
exports.success = success;
const error = (message) => chalk_1.default.redBright('✖ error ') + message;
exports.error = error;
const note = (message) => chalk_1.default.gray('  → ' + message);
exports.note = note;
exports.default = { info: exports.info, warn: exports.warn, pending: exports.pending, success: exports.success, error: exports.error, note: exports.note };
